File: smart_bot.cpp
#include <time.h>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <queue>
#include <set>
#include <algorithm>

#include "helper-package/bot.h"
#include "helper-package/constants.h"

using namespace std;

fstream fs;

class SmartBot : public Bot {
public:
    SmartBot(int minPollen, float ratio) : Bot("SmartBot"), minPollen(minPollen), ratio(ratio) { }

protected:
    vector<Action> getMoves(int id, const Map& curMap) {
        vector<Action> moves;

        // Move to closest flower or queen bee if bee has pollen.
        for (auto beeCell = beeCells.begin(); beeCell != beeCells.end(); ++beeCell) {
            shared_ptr<Bee> bee = (*beeCell)->getBee();
            Position pos = (*beeCell)->getPosition();
            if (bee->pollen >= minPollen * bee->count) {
                if (!isBesideHiveOrQueen(bee->pos)) {
                    Map::Path minPath = curMap.getPath(pos, queenBee->pos);
                    random_shuffle(hiveCells.begin(), hiveCells.end());
                    for (auto hiveCell = hiveCells.begin(); hiveCell != hiveCells.end(); ++hiveCell) {
                        Map::Path path = curMap.getPath(pos, (*hiveCell)->getPosition());
                        if (path.distance < minPath.distance) {
                            minPath = path;
                        }
                    }
                    moves.push_back(
                            Action(actionType::MOVE, pos, minPath.move, minPath.move - 1 /* Faces move direction. */));
                }
            } else {
                Cell* firstFlower = flowerCells[0];
                Map::Path minPath = curMap.getPath(pos, firstFlower->getPosition());
                random_shuffle(flowerCells.begin(), flowerCells.end());
                for (auto flowerCell = flowerCells.begin() + 1; flowerCell != flowerCells.end(); ++flowerCell) {
                    Map::Path path = curMap.getPath(pos, (*flowerCell)->getPosition());
                    if (path.distance < minPath.distance) {
                        minPath = path;
                    }
                }
                moves.push_back(
                        Action(actionType::MOVE, pos, minPath.move, minPath.move - 1 /* Faces move direction. */));
            }
        }

        int pollenNotUsed = queenBee->pollen;
        Position queenPos = queenBee->pos;

        // Check if queen is beside an enemy, and move away from enemies if so.
        set<Move> availableMoves = { move::RIGHT, move::DOWN, move::LEFT, move::UP };
        if (hasEnemy(getBoundedPos(queenPos.x + 1, queenPos.y))) {
            availableMoves.erase(move::RIGHT);
            availableMoves.erase(move::DOWN);
            availableMoves.erase(move::UP);
        }
        if (hasEnemy(getBoundedPos(queenPos.x + 1, queenPos.y + 1))) {
            availableMoves.erase(move::RIGHT);
            availableMoves.erase(move::DOWN);
        }
        if (hasEnemy(getBoundedPos(queenPos.x + 1, queenPos.y - 1))) {
            availableMoves.erase(move::RIGHT);
            availableMoves.erase(move::UP);
        }
        if (hasEnemy(getBoundedPos(queenPos.x - 1, queenPos.y))) {
            availableMoves.erase(move::LEFT);
            availableMoves.erase(move::DOWN);
            availableMoves.erase(move::UP);
        }
        if (hasEnemy(getBoundedPos(queenPos.x - 1, queenPos.y + 1))) {
            availableMoves.erase(move::LEFT);
            availableMoves.erase(move::DOWN);
        }
        if (hasEnemy(getBoundedPos(queenPos.x - 1, queenPos.y - 1))) {
            availableMoves.erase(move::LEFT);
            availableMoves.erase(move::UP);
        }
        if (hasEnemy(getBoundedPos(queenPos.x, queenPos.y + 1))) {
            availableMoves.erase(move::LEFT);
            availableMoves.erase(move::RIGHT);
            availableMoves.erase(move::UP);
        }
        if (hasEnemy(getBoundedPos(queenPos.x, queenPos.y - 1))) {
            availableMoves.erase(move::LEFT);
            availableMoves.erase(move::RIGHT);
            availableMoves.erase(move::DOWN);
        }
        if (availableMoves.size() < 4) {
            // Case when you won't escape an enemy no matter where you move.
            if (availableMoves.size() == 0) {
                moves.push_back(Action(actionType::MOVE_QUEEN, (rand() % 4) + 1));
            } else {
                vector<Move> shuffledMoves(availableMoves.begin(), availableMoves.end());
                random_shuffle(shuffledMoves.begin(), shuffledMoves.end());
                moves.push_back(Action(actionType::MOVE_QUEEN, *shuffledMoves.begin()));
            }
        } else {
            int posPotency = curMap.map[queenPos.y][queenPos.x]->getPotency();
            // Create hive on spot
            if (totalBees > ratio * hiveCells.size() && posPotency == 0) {
                pollenNotUsed -= HIVE_POLLEN_AMOUNT;
                moves.push_back(Action(actionType::CREATE_HIVE));
            }

            // If queen is on hive or flower, move to an open spot.
            if (posPotency > 0 || curMap.map[queenPos.y][queenPos.x]->getOwnerId() == id) {
                Move movements[] = {move::RIGHT, move::DOWN, move::LEFT, move::UP};
                random_shuffle(begin(movements), end(movements));

                bool moved = false;
                for (int i = 0; i < 4; ++i) {
                    if (movements[i] == move::RIGHT &&
                        isEmpty(getBoundedPos(queenPos.x + 1, queenPos.y)) &&
                        !hasEnemy(getBoundedPos(queenPos.x + 1, queenPos.y))) {
                        moved = true;
                        moves.push_back(Action(actionType::MOVE_QUEEN, move::RIGHT));
                        break;
                    } else if (movements[i] == move::DOWN &&
                               isEmpty(getBoundedPos(queenPos.x, queenPos.y + 1)) &&
                               !hasEnemy(getBoundedPos(queenPos.x, queenPos.y + 1))) {
                        moved = true;
                        moves.push_back(Action(actionType::MOVE_QUEEN, move::DOWN));
                        break;
                    } else if (movements[i] == move::LEFT &&
                               isEmpty(getBoundedPos(queenPos.x - 1, queenPos.y)) &&
                               !hasEnemy(getBoundedPos(queenPos.x - 1, queenPos.y))) {
                        moved = true;
                        moves.push_back(Action(actionType::MOVE_QUEEN, move::LEFT));
                        break;
                    } else if (movements[i] == move::UP &&
                               isEmpty(getBoundedPos(queenPos.x, queenPos.y - 1)) &&
                               !hasEnemy(getBoundedPos(queenPos.x, queenPos.y - 1))) {
                        moved = true;
                        moves.push_back(Action(actionType::MOVE_QUEEN, move::UP));
                        break;
                    }
                }
                // Everywhere is blocked so just move to a random movement that is not STAY.
                if (!moved) {
                    moves.push_back(Action(actionType::MOVE_QUEEN, (rand() % 4) + 1));
                }
            }
        }

        if (hiveCells.size() > 0) {
            int totalBeesPerHive = (pollenNotUsed / BEE_POLLEN_AMOUNT) / hiveCells.size();
            for (auto &hiveCell : hiveCells) {
                if (totalBeesPerHive > 100) {
                    moves.push_back(
                            Action(
                                    actionType::SPAWN,
                                    hiveCell->getPosition(),
                                    MAX_BEES
                            )
                    );
                } else {
                    moves.push_back(
                            Action(
                                    actionType::SPAWN,
                                    hiveCell->getPosition(),
                                    totalBeesPerHive
                            )
                    );
                }
            }
        }

        return moves;
    }

private:
    const int minPollen;
    const float ratio;

    // Detects when a spot has no hive or flower on it.
    bool isEmpty(Position pos) {
        return curMap->map[pos.y][pos.x]->getPotency() == 0 && curMap->map[pos.y][pos.x]->getOwnerId() == NEUTRAL_ID;
    }
};

int main() {
    srand(time(0));

    fs.open("smart_bot_log.txt", fstream::out | fstream::trunc);

    cout.sync_with_stdio(false);

    SmartBot smartBot = SmartBot(1, 30.0f);

    smartBot.run();

    fs.close();
}

File: smart-bot.js
const CONSTANTS = require('./helper-package-js/constants');
const { ACTION, FACE, FLOWER, MOVE } = require('./helper-package-js/constants');

const { Action } = require('./helper-package-js/action');
const { Bot } = require('./helper-package-js/bot');
const { Position } = require('./helper-package-js/position');

const BEES_TO_HIVE_RATIO = 30;

class SmartBot extends Bot {
    constructor() {
        super("SmartBot");
    }

    getActions() {
        const actions = [];

        for (let beeCell of this.beeCells) {
            const bee = beeCell.bee;
            const pos = beeCell.pos;
            if (bee.pollen >= bee.count) {
                // If the bees have pollen, move to closest queen or hive.
                if (!this.isBesideHiveOrQueen(pos)) {
                    let minPath = this.getMinPath(pos, this.queenBee.pos);

                    // Randomizes hive priorities if same distance.
                    this.shuffle(this.hiveCells);

                    for (let hiveCell of this.hiveCells) {
                        let path = this.getMinPath(pos, hiveCell.pos);
                        if (path.distance < minPath.distance) {
                            minPath = path;
                        }
                    }
                    actions.push(new Action({
                        type: ACTION.MOVE,
                        pos: pos,
                        move: minPath.move,
                        face: minPath.move - 1 /* Faces move direction. */,
                    }));
                }
            } else {
                // Randomizes flower priorities if same distance.
                this.shuffle(this.flowerCells);

                // Move to closest flower to collect pollen.
                const firstFlower = this.flowerCells[0];
                let minPath = this.getMinPath(pos, firstFlower.pos);

                for (let i = 1; i < this.flowerCells.length; ++i) {
                    const path = this.getMinPath(pos, this.flowerCells[i].pos);
                    if (path.distance < minPath.distance) {
                        minPath = path;
                    }
                }
                actions.push(new Action({
                    type: ACTION.MOVE,
                    pos: pos,
                    move: minPath.move,
                    face: minPath.move - 1 /* Faces move direction. */,
                }));
            }
        }

        const queenPos = this.queenBee.pos;
        let pollenNotUsed = this.queenBee.pollen;
        let posPotency = this.currMap.map[queenPos.y][queenPos.x].potency;

        // Check for an enemy at each position beside the queen.
        const topLeftEnemy = this.hasEnemy(this.getBoundedPos(queenPos.x - 1, queenPos.y - 1));
        const topEnemy = this.hasEnemy(this.getBoundedPos(queenPos.x, queenPos.y - 1));
        const topRightEnemy = this.hasEnemy(this.getBoundedPos(queenPos.x + 1, queenPos.y - 1));
        const leftEnemy = this.hasEnemy(this.getBoundedPos(queenPos.x - 1, queenPos.y));
        const rightEnemy = this.hasEnemy(this.getBoundedPos(queenPos.x + 1, queenPos.y));
        const botLeftEnemy = this.hasEnemy(this.getBoundedPos(queenPos.x - 1, queenPos.y + 1));
        const botEnemy = this.hasEnemy(this.getBoundedPos(queenPos.x, queenPos.y + 1));
        const botRightEnemy = this.hasEnemy(this.getBoundedPos(queenPos.x + 1, queenPos.y + 1));

        // Add moves that will move the queen away from potential threats.
        const availableMoves = [];
        if (!topEnemy && !botEnemy && !topLeftEnemy && !leftEnemy && !botLeftEnemy) {
            availableMoves.push(MOVE.LEFT);
        }
        if (!leftEnemy && !rightEnemy && !botLeftEnemy && !botEnemy && !botRightEnemy) {
            availableMoves.push(MOVE.DOWN);
        }
        if (!topEnemy && !botEnemy && !topRightEnemy && !rightEnemy && !botRightEnemy) {
            availableMoves.push(MOVE.RIGHT);
        }
        if (!leftEnemy && !rightEnemy && !topLeftEnemy && !topEnemy && !topRightEnemy) {
            availableMoves.push(MOVE.UP);
        }

        // If available moves is less than 4, then that means the queen is in threat of an enemy.
        // We should move it then.
        if (availableMoves.length < 4) {
            // Case when you won't escape an enemy no matter where you move.
            if (availableMoves.length == 0) {
                availableMoves.push(Math.floor(Math.random() % 4) + 1);
            }
            const move = availableMoves[Math.floor(Math.random() * availableMoves.length)];
            actions.push(new Action({
                type: ACTION.MOVE_QUEEN,
                move: move,
                face: move - 1,
            }));
        } else {
            // No threat, so proceed to expand normally.

            // Create hive if the bees to hive ratio is higher than what we want.
            if (this.totalBees > BEES_TO_HIVE_RATIO * this.hiveCells.length && posPotency == FLOWER.NONE) {
                pollenNotUsed -= CONSTANTS.HIVE_POLLEN_AMOUNT;
                actions.push(new Action({
                    type: ACTION.CREATE_HIVE,
                }));
            }

            // If queen is on a hive or flower, randomly move to an open spot.
            if (!this.isEmpty(queenPos)) {
                const movements = [MOVE.RIGHT, MOVE.DOWN, MOVE.LEFT, MOVE.UP];
                this.shuffle(movements);

                let moved = false;
                for (let i = 0; i < 4; ++i) {
                    if (movements[i] == MOVE.RIGHT &&
                        this.isEmpty(this.getBoundedPos(queenPos.x + 1, queenPos.y)) &&
                        !this.hasEnemy(this.getBoundedPos(queenPos.x + 1, queenPos.y))) {
                        moved = true;
                        actions.push(new Action({
                            type: ACTION.MOVE_QUEEN,
                            move: MOVE.RIGHT,
                            face: FACE.RIGHT,
                        }));
                        break;
                    } else if (movements[i] == MOVE.DOWN &&
                        this.isEmpty(this.getBoundedPos(queenPos.x, queenPos.y + 1)) &&
                        !this.hasEnemy(this.getBoundedPos(queenPos.x, queenPos.y + 1))) {
                        moved = true;
                        actions.push(new Action({
                            type: ACTION.MOVE_QUEEN,
                            move: MOVE.DOWN,
                            face: FACE.DOWN,
                        }));
                        break;
                    } else if (movements[i] == MOVE.LEFT &&
                        this.isEmpty(this.getBoundedPos(queenPos.x - 1, queenPos.y)) &&
                        !this.hasEnemy(this.getBoundedPos(queenPos.x - 1, queenPos.y))) {
                        moved = true;
                        actions.push(new Action({
                            type: ACTION.MOVE_QUEEN,
                            move: MOVE.LEFT,
                            face: FACE.LEFT,
                        }));
                        break;
                    } else if (movements[i] == MOVE.UP &&
                        this.isEmpty(this.getBoundedPos(queenPos.x, queenPos.y - 1)) &&
                        !this.hasEnemy(this.getBoundedPos(queenPos.x, queenPos.y - 1))) {
                        moved = true;
                        actions.push(new Action({
                            type: ACTION.MOVE_QUEEN,
                            move: MOVE.UP,
                            face: FACE.UP,
                        }));
                        break;
                    }
                }

                // Everywhere is blocked, so just move randomly.
                if (!moved) {
                    const move = Math.floor(Math.random() * 4) + 1;
                    actions.push(new Action({
                        type: ACTION.MOVE_QUEEN,
                        move: move,
                        face: move - 1,
                    }));
                }
            }
        }

        if (this.hiveCells.length > 0) {
            const totalBeesPerHive =
                Math.floor(Math.floor(pollenNotUsed / CONSTANTS.BEE_POLLEN_AMOUNT) / this.hiveCells.length);
            for (let hiveCell of this.hiveCells) {
                if (totalBeesPerHive > CONSTANTS.MAX_BEES) {
                    actions.push(new Action({
                        type: ACTION.SPAWN,
                        pos: hiveCell.pos,
                        amount: CONSTANTS.MAX_BEES,
                    }));
                } else {
                    actions.push(new Action({
                        type: ACTION.SPAWN,
                        pos: hiveCell.pos,
                        amount: totalBeesPerHive,
                    }));
                }
            }
        }

        return actions;
    }

    isEmpty(pos) {
        return this.currMap.map[pos.y][pos.x].potency == FLOWER.NONE &&
            this.currMap.map[pos.y][pos.x].ownerId == CONSTANTS.NEUTRAL_ID;
    }

    // Fisher-Yates shuffle algorithm.
    shuffle(array) {
        for (let i = array.length - 1; i > 0; --i) {
            const randomInd = Math.floor(Math.random() * (i + 1));
            const temp = array[i];
            array[i] = array[randomInd];
            array[randomInd] = temp;
        }
    }
}

(async function main() {
    const smartBot = new SmartBot();
    await smartBot.run();
})();
